// libweb/src/components/filer/init.js
import store from './store';
import { ServerValues } from "../../utility/ServerValues";

/**
 * Install
 *
 * options = {  
 *   store, // required
 *
 *   // not required params
 *   headers: {'Authorization': 'Bearer ...'}, // add header
 * OR
 *   headers: {'X-CSRF-TOKEN': 'token'}, // overwrite default header Axios
 *   baseUrl: 'http://my_url:80/file-manager/', // overwrite base url Axios
 *   windowsConfig: 2, // overwrite config
 *   lang: 'de',
 *   translation: {
 *     name: de,
 *     content: {
 *       about: 'Über',
 *       back: 'Zurück',
 *       ... see lang file structure
 *     },
 *   },
 * }
 *
 * @param options
 */
export function install_filer() {
    const serverValues = new ServerValues().getValues();
    var headers = {};
    headers[serverValues['xsrf-header-name']] = serverValues['xsrf-token'];
    var options = {
        store: Vue.prototype.$store,
        baseUrl: '/api/filemanager/lfm/',
        windowsConfig: 2,
        lang: serverValues['culture'],
        headers: headers,
    };
    if (!options.store) console.error('Please provide a store!!');
    options.store.registerModule('fm', store);
    options.store.commit('fm/settings/manualSettings', options);
}
