// libweb/src/components/html-editor/blocks/image-url/ImageUrl.js
import Block from '../Block';
export default class ImageUrl extends Block {
    getType() {
        return 'image-url';
    }
    getEmptyData() {
        var res = {
            url: '',
            caption: '',
        };
        return Object.assign({}, res, this.getEmptySettingsInData());;
    }
    getEmptySettingsInData() {
        return {
            width: '100%',
            minWidth: '300px',
            maxWidth: '900px',
            height: 'auto',
            minHeight: '',
            maxHeight: ''
        };
    }
    isDataValid(data) {
        var keys = Object.keys(data);
        return keys.length === 8 && !_.isNil(data.url) && !_.isNil(data.caption);
    }
    getEmptyConfig() {
        return {
            validExtensions: ['jpg', 'jpeg', 'gif', 'png', 'svg']
        };
    }
    mergeData(dest, src) {
        if (!_.isNil(src.url)) {
            dest.url = src.url;
        }
        if (!_.isNil(src.caption)) {
            dest.caption = src.caption;
        }
        if (!_.isNil(src.width)) {
            dest.width = src.width;
        }
        if (!_.isNil(src.minWidth)) {
            dest.minWidth = src.minWidth;
        }
        if (!_.isNil(src.maxWidth)) {
            dest.maxWidth = src.maxWidth;
        }
        if (!_.isNil(src.height)) {
            dest.height = src.height;
        }
        if (!_.isNil(src.minHeight)) {
            dest.minHeight = src.minHeight;
        }
        if (!_.isNil(src.maxHeight)) {
            dest.maxHeight = src.maxHeight;
        }
    }
}
