// libweb/src/components/file-select/acceptToFileTypes.js
/**
 * convert Html5 accept string to array of separate accepts
 * @param {String} accept example: image/\*, audio/\*, video/\*, ....
 */
export function acceptToFileTypes(accept) {
    const __image = {
        any: 'image/*',
        jpg: 'image/jpg,image/jpeg',
        png: 'image/png',
        gif: 'image/gif',
        bmp: 'image/bmp,image/x-bmp,image/vnd.wap.wbmp',
    };
    const __audio = {
        any: 'audio/*',
        mp3: 'audio/mpeg',
        midi: 'audio/midi',
        amr: 'audio/amr,audio/amr-wb',
        wav: 'audio/wav,audio/x-wav'
    };
    const __video = {
        any: 'video/*',
        '3gp': 'audio/3gpp',
        mp4: 'audio/mp4',
    };
    const allAsArray = (o) => {
        const res = [];
        Object.keys(o).forEach(key => {
            if (key === 'any') {
                return;
            }
            o[key].split(',').forEach(a => {
                res.push(a.trim());
            });
        });
        return res;
    };
    if (accept === undefined || accept === null || accept.length === 0) {
        return [];
    }
    const array = accept.toString().toLowerCase().trim().split(",");
    const res = [];
    array.forEach(a => {
        const type = a.toString().toLowerCase().trim();
        let k = [];
        if (type === "image/*") {
            k = allAsArray(__image);
        } else if (type === "audio/*") {
            k = allAsArray(__audio);
        } else if (type === "video/*") {
            k = allAsArray(__video);
        } else {
            k = [a.toLowerCase().trim()];
        }
        k.forEach(x => {
            res.push(x);
        });
    });
    return res;
}
