// libweb/src/plugins/responsive/responsiveStoreModule.js
var responsiveStoreModule = {
    state: {
        // <= 575.98
        xs: false,
        // >= 576 && <= 767.98
        sm: false,
        smAndUp: false,
        smAndDown: false,
        // >= 768 && <= 991.98
        md: false,
        mdAndUp: false,
        mdAndDown: false,
        // >= 992 && <= 1199.98
        lg: false,
        lgAndUp: false,
        lgAndDown: false,
        // >= 1200
        xl: false,
    },
    mutations: {
        /**
         * 
         * @param {Object} state current state
         * @param {Object} args arguments
         * @param {String} args.variable: xs, sm, md, lg, xl
         * @param {Boolean} args.isMatch
         */
        setBreakpoints(state, args) {
            state[args.variable] = args.isMatch;
            state.smAndUp = !state.xs;
            state.smAndDown = state.xs || state.sm;
            state.mdAndUp = !state.xs && !state.sm;
            state.mdAndDown = state.xs || state.sm || state.md;
            state.lgAndUp = !state.xs && !state.sm && !state.md;
            state.lgAndDown = state.xs || state.sm || state.md || state.lg;
        },
    }
}

export { responsiveStoreModule }
