// libweb/src/components/html-editor/blockEditorRegistry.js
// text
import text from './blocks/text/text.vue';
import Text from './blocks/text/Text';
import textLoc from './blocks/text/text.i18n.json';
// paragraph
import paragraph from './blocks/paragraph/paragraph.vue';
import Paragraph from './blocks/paragraph/Paragraph';
import paragraphLoc from './blocks/paragraph/paragraph.i18n.json';
// header
import header from './blocks/header/header.vue';
import Header from './blocks/header/Header';
import headerLoc from './blocks/header/header.i18n.json';
// delimiter
import delimiter from './blocks/delimiter/delimiter.vue';
import Delimiter from './blocks/delimiter/Delimiter';
import delimiterLoc from './blocks/delimiter/delimiter.i18n.json';
// image-url
import imageUrl from './blocks/image-url/image-url.vue';
import ImageUrl from './blocks/image-url/ImageUrl';
import imageUrlLoc from './blocks/image-url/image-url.i18n.json';
// youtube
import youtube from './blocks/youtube/youtube.vue';
import Youtube from './blocks/youtube/Youtube';
import youtubeLoc from './blocks/youtube/youtube.i18n.json';
// aparat
import aparat from './blocks/aparat/aparat.vue';
import Aparat from './blocks/aparat/Aparat';
import aparatLoc from './blocks/aparat/aparat.i18n.json';
// raw html
import raw from './blocks/raw-html/rawHtml.vue';
import Raw from './blocks/raw-html/RawHtml';
import rawLoc from './blocks/raw-html/rawHtml.i18n.json';
// images-url
import imagesUrl from './blocks/images-url/images-url.vue';
import ImagesUrl from './blocks/images-url/ImagesUrl';
import imagesUrlLoc from './blocks/images-url/images-url.i18n.json';
// other imports
import Block from './blocks/Block';
class BlockEditorRegistry {
    constructor() {
        this.modules = {};
    }
    getModules() {
        return this.modules;
    }
    /**
     * Register block
     * @param {VueComponent} vueComponent vue component of the block 
     * @param {Block} descriptor block descriptor 
     * @param {Object} loc localization file instance 
     */
    register(vueComponent, descriptor, loc) {
        var type = descriptor.getType();
        var t = type.startsWith('block-') ? type : `block-${type}`;
        this.modules[t] = {
            descriptor: descriptor,
            loc: loc
        };
        Vue.component(t, vueComponent);
    }
    /**
     * @returns {Array<String>}
     */
    getTypes() {
        return Object.keys(this.modules);
    }
    getLoc() {
        var res = {};
        this.getTypes().forEach(type => {
            var loc = this.modules[type].loc;
            Object.keys(loc).forEach(culture => {
                if (_.isNil(res[culture])) {
                    res[culture] = {};
                }
                var o = loc[culture][type];
                res[culture][type] = {
                    'type': o.type,
                    'description': o.description
                };
            });
        });
        return res;
    }
}
const blockEditorRegistry = new BlockEditorRegistry();

//register default modules
blockEditorRegistry.register(text, new Text(), textLoc);
blockEditorRegistry.register(header, new Header(), headerLoc);
blockEditorRegistry.register(paragraph, new Paragraph(), paragraphLoc);
blockEditorRegistry.register(delimiter, new Delimiter(), delimiterLoc);
blockEditorRegistry.register(imageUrl, new ImageUrl(), imageUrlLoc);
blockEditorRegistry.register(youtube, new Youtube(), youtubeLoc);
blockEditorRegistry.register(aparat, new Aparat(), aparatLoc);
blockEditorRegistry.register(raw, new Raw(), rawLoc);
blockEditorRegistry.register(imagesUrl, new ImagesUrl(), imagesUrlLoc);

export { blockEditorRegistry }
