// libweb/src/utility/guid.js
class Guid {
    /**
     * Creates a new guid
     * @param {Boolean} hyphen if true then the guid will be like aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee otherwise there will be no hyphen
     * @returns {String}
     */
    create(hyphen) {
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
        }
        if (hyphen === undefined || hyphen === null) {
            return `${s4()}${s4()}${s4()}${s4()}${s4()}${s4()}${s4()}${s4()}`;
        }
        return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
    }
    /**
     * Returns a new random id using create(false)
     * @returns {String}
     */
    newId() {
        return `e_${this.create()}`;
    }
    /**
     * Empty value for guid : '00000000-0000-0000-0000-000000000000'
     * @returns {String}
     */
    empty() {
        return '00000000-0000-0000-0000-000000000000';
    }
    /**
     * Empty value for guid : '00000000000000000000000000000000'
     * @returns {String}
     */
    emptyNoHyphen() {
        return '00000000000000000000000000000000';
    }
}
var guid = new Guid();
export { guid }
